torch/nn/modules/rnn.py

```diff
@@ -396,6 +398,65 @@ def __init__(self, *args, **kwargs):
             raise ValueError("Unknown nonlinearity '{}'".format(self.nonlinearity))
         super(RNN, self).__init__(mode, *args, **kwargs)
 
+    @overload
+    @torch._jit_internal._overload_method  # noqa: F811
+    def forward(self, input, hx=None):  # noqa: F811
+        # type: (Tensor, Optional[Tensor]) -> Tuple[Tensor, Tensor]
+        pass
+
+    @overload
+    @torch._jit_internal._overload_method  # noqa: F811
+    def forward(self, input, hx=None):  # noqa: F811
+        # type: (PackedSequence, Optional[Tensor]) -> Tuple[PackedSequence, Tensor]
+        pass
+
+    def forward(self, input, hx=None):  # noqa: F811
+        orig_input = input
+        # xxx: isinstance check needs to be in conditional for TorchScript to compile
+        if isinstance(orig_input, PackedSequence):
+            input, batch_sizes, sorted_indices, unsorted_indices = input
+            max_batch_size = batch_sizes[0]
+            max_batch_size = int(max_batch_size)
+        else:
+            batch_sizes = None
+            max_batch_size = input.size(0) if self.batch_first else input.size(1)
+            sorted_indices = None
+            unsorted_indices = None
+
+        if hx is None:
+            num_directions = 2 if self.bidirectional else 1
+            hx = torch.zeros(self.num_layers * num_directions,
+                             max_batch_size, self.hidden_size,
+                             dtype=input.dtype, device=input.device)
+        else:
+            # Each batch of the hidden state should match the input sequence that
+            # the user believes he/she is passing in.
+            hx = self.permute_hidden(hx, sorted_indices)
+
+        self.check_forward_args(input, hx, batch_sizes)
+        if self.mode == "RNN_TANH":
```

Comment:
Can you initialize `_VF.rnn_tanh` / `_VF.rnn_relu` in the initialization as `self.activation` ? And then do the same with `nn.GRU`? I think if you did that you could have `RNN` and `GRU` extend from the same base class that defines the forward method, and then just separately define `self.activation` in the `__init__`

It's possible that `_overload_method` doesn't work yet with inheritance and that doesn't work though...

Reply:
I tried the overload with inheritance and it looks like that doesn't work.
https://gist.github.com/you74674/48dd48ba2b21cc5af40aa512a43b4f32